File: JavaDataManager/src/joo/databaseManager/SQL/Select.java
package joo.databaseManager.SQL;

import java.util.ArrayList;
import java.util.List;

import joo.databaseManager.SQL.OrderBy.ORDER;
import joo.databaseManager.SQL.Where.Operator;
import joo.databaseManager.connection.DataBaseConnection;
import joo.databaseManager.type.DatabaseType;
import joo.databaseManager.type.FieldInfo;

public class Select {

	private List<String> fields;
    private List<String> froms;
    private List<String> where;
    private int limitStart;
    private int limitLength;
    private OrderBy orderBy;
    private List<Object> whereParameters;
    private String groupBy;
    
	@SuppressWarnings("unchecked")
	public Select(DatabaseType type) {
		groupBy=" group by "+type.getTableName()+".id ";
		fields = new ArrayList<String>();
        froms = new ArrayList<String>();
        where = new ArrayList<String>();
        whereParameters=new ArrayList<Object>();
        
        addFrom(type.getTableName());
        //groupBy = " group by " + type.TableName + ".id";
        OrderBy orderBy = new OrderBy(ORDER.ASC);
        orderBy.addItem(type.getModelClass(), "ID");

        for (FieldInfo field: type.getFields())
        {
            addField(type.getTableName() + "." + field.getName());
        }
	}
	
	public void addField(String name)
    {
        for (String item : fields) {
			if(item==name){
				return;
			}
		}
        fields.add(name);
    }

    private void addWhere(String item)
    {
        where.add(item);
    }
    
    public void addFrom(String table)
    {
        for (String item:froms) {
			if(item==table){
				return;
			}
		}
        froms.add(table);
    }

    public boolean hasFrom()
    {
        return froms.size() > 0;
    }

    public boolean hasField()
    {
        return fields.size() > 0;
    }

    public boolean hasWhere()
    {
        return where.size() > 0;
    }

	public int getLimitStart() {
		return limitStart;
	}

	public void setLimitStart(int limitStart) {
		this.limitStart = limitStart;
	}

	public int getLimitLength() {
		return limitLength;
	}

	public void setLimitLength(int limitLength) {
		this.limitLength = limitLength;
	}

	public OrderBy getOrderBy() {
		return orderBy;
	}

	public void setOrderBy(OrderBy orderBy) {
		this.orderBy = orderBy;
	}
	public List<Object> getParameters(){
		return whereParameters;
	}

	public void addWhere(Where where, DataBaseConnection connection) {
        if(where==null){
        	return;
        }
		List<Object> items=where.getItems();
        for (int i = 0; i < items.size(); i++)
        {
            Object item = items.get(i);
            if (item instanceof FieldInfo)
            {
                if (items.size() > i + 1 && items.get(i + 1) instanceof Operator && ((Operator)items.get(i + 1)) == Operator.CONTAINS)
                {
                    continue;
                }
                FieldInfo info=(FieldInfo)item;
                DatabaseType type=new DatabaseType(info.getOwnerClazz());
                String tableName=type.getTableName();
                if(tableName!=null && tableName!=""){
                	addFrom(tableName);
                }
                addWhere(type.getTableName() + "." + info.getName());
            }
            else
            {
                if (item instanceof Operator)
                {
                    switch ((Operator)item)
                    {
                        case AND:
                        case OR:
                        case OPEN_PARENTHESIS:
                        case CLOSE_PARENTHESIS:
                        case DIFFERENT:
                        case EQUAL:
                        case LESS_EQUAL:
                        case MINOR:
                        case MORE:
                        case MORE_EQUAL:
                            addWhere(connection.convertOperatorToString((Operator)item));
                            break;
                        case CONTAINS:
                            {
                                if (!(items.get(i-1) instanceof FieldInfo ))
                                {
                                    throw new IllegalAccessError("Where is not valid. Contains syntax invalid");
                                }
                                
                                if (!(items.get(i+1) instanceof String ) && !(items.get(i+1) instanceof FieldInfo ))
                                {
                                    throw new IllegalAccessError("Where is not valid. Contains syntax invalid");
                                }
                                String aux="";
                                FieldInfo info=(FieldInfo)items.get(i-1);
                                DatabaseType type=new DatabaseType(info.getOwnerClazz());
                                addFrom(type.getTableName());
                                aux += type.getTableName() + "." + info.getName();
                                
                                
                                aux += " " + connection.convertOperatorToString((Operator)item) + " ";

                                if (items.get(i+1) instanceof FieldInfo)
                                {
                                    info=(FieldInfo)items.get(i+1);
                                    type=new DatabaseType(info.getOwnerClazz());
                                    addFrom(type.getTableName());
                                    aux += type.getTableName() + "." + info.getName();
                                }
                                else
                                {
                                    aux += "'%"+items.get(i+1).toString()+"%'";
                                }
                                
                                addWhere(aux);
                                i++;
                            }
                            break;
                    }
                }
                else
                {
                    addWhere("?");
                    whereParameters.add(item);
                }
            }
        }
	}
	
	@Override
	public String toString(){
		String ret="select ";
		for (int i = 0; i < fields.size(); i++) {
			ret+=fields.get(i);
			if(i<fields.size()-1){
				ret+=", ";
			}
		}
		ret+=" from ";
		for (int i = 0; i < froms.size(); i++) {
			ret+="`"+froms.get(i)+"`";
			if(i<froms.size()-1){
				ret+=", ";
			}
		}
		
		if(where.size()>0){
			ret+=" where ";
			for (int i = 0; i < where.size(); i++) {
				ret+=where.get(i)+" ";
			}
		}
		
		ret+=groupBy;
		
		if(orderBy!=null && orderBy.getItems().size()>0){
			ret+="order by ";
			List<Object[]> items=orderBy.getItems();
			for (int i = 0; i < items.size(); i++) {
				DatabaseType type=(DatabaseType)items.get(i)[0];
				String field=(String)items.get(i)[1];
				ret+=type.getTableName()+"."+field;
				if(i<items.size()-1){
					ret+=", ";
				}
			}
			ret+=" "+orderBy.getOrder();
		}
		if (limitStart >= 0 && limitLength > 0)
        {
            ret+=" LIMIT " + limitStart + " , " + limitLength;
        }
		
        return ret;
	}

}

File: NetDataManager/Database/Structs/Select.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Database.Connections;
using System.Reflection;
using Database.Types;

namespace Database.Structs
{
    public class Select
    {
        private List<string> fields;
        private List<string> froms;
        private List<string> where;
        //private string groupBy;

        public Select()
        {
            fields = new List<string>();
            froms = new List<string>();
            where = new List<string>();
            //groupBy = string.Empty;
        }
        public Select(DatabaseType type,bool onDemand)
        {
            fields = new List<string>();
            froms = new List<string>();
            where = new List<string>();
            AddFrom(type.TableName);
            //groupBy = " group by " + type.TableName + ".id";
            OrderBy = new OrderBy();
            OrderBy.AddItem(type.Type, "ID");

            foreach (var field in type.DataBaseProperties)
            {
                if (onDemand)
                {
                    if (!field.IsOnDemandField)
                    {
                        AddField(type.TableName + "." + field.Attribute.Name);
                    }
                }
                else
                {
                    AddField(type.TableName + "." + field.Attribute.Name);
                }
            }
        }

        public override string ToString()
        {
            StringBuilder select = new StringBuilder();
            select.Append("select ");
            foreach (string item in fields)
            {
                select.Append(item);
                if (fields.IndexOf(item) != fields.Count - 1)
                {
                    select.Append(", ");
                }
            }

            select.Append(" from ");

            foreach (string item in froms)
            {
                select.Append(item);
                if (froms.IndexOf(item) != froms.Count - 1)
                {
                    select.Append(", ");
                }
            }

            if (where.Count > 0)
            {
                select.Append(" where ");
                foreach (string item in where)
                {
                    select.Append(item + " ");
                }
            }

            if (OrderBy!=null)
            {
                select.Append(" order by ");
                bool more = false;

                if (OrderBy.Items.Count > 0)
                {
                    DatabaseType type = TypesManager.TypeOf(OrderBy.Items[0][0] as Type);
                    foreach (Object[] item in OrderBy.Items)
                    {
                        if (more)
                        {
                            select.Append(", ");
                        }
                     
                        foreach (var property in type.DataBaseProperties)
                        {
                            if (property.Property.Name == item[1] as string)
                            {
                                select.Append(type.TableName + "." + property.Attribute.Name);
                                more = true;
                                break;
                            }
                        }
                    }
                }
            }
            
            if (LimitStart > 0 && LimitLength > 0)
            {
                select.Append(" LIMIT " + LimitStart + " , " + LimitLength + "");
            }
            return select.ToString();
        }

        public void AddField(string name)
        {
            foreach (string item in fields)
            {
                if (name == item)
                {
                    return;
                }
            }
            fields.Add(name);
        }

        private void AddWhere(string item)
        {
            where.Add(item);
        }

        public Dictionary<string,object> AddWhere(Where where,DataBaseConnection connection)
        {
            Dictionary<string, object> ret = new Dictionary<string, object>();
            int parameterCount = 1;
            for (int i = 0; i < where.Items.Count; i++)
            {
                object item = where.Items[i];
                // § //if (item is PropertyInfo)
                if (item is DatabasePropertyInfo)
                {
                    if (where.Items.Count > i + 1 && where.Items[i + 1] is Operator && ((Operator)where.Items[i + 1]) == Operator.CONTAINS)
                    {
                        continue;
                    }
                    // § // DatabaseType tempType = TypesManager.TypeOf((item as PropertyInfo).ReflectedType);
                    // § // DatabaseFieldInfo databaseInfo = tempType.GetPropertyInfo((item as PropertyInfo).Name) as DatabaseFieldInfo;

                    var dtInfo = (item as DatabaseFieldInfo);
                    AddFrom(dtInfo.TableName);
                    AddWhere(dtInfo.TableName + "." + dtInfo.Attribute.Name);
                }
                else
                {
                    if (item is Operator)
                    {
                        switch ((Operator)item)
                        {
                            case Operator.AND:
                            case Operator.OR:
                            case Operator.OPEN_PARENTHESIS:
                            case Operator.CLOSE_PARENTHESIS:
                            case Operator.DIFFERENT:
                            case Operator.EQUAL:
                            case Operator.LESS_EQUAL:
                            case Operator.MINOR:
                            case Operator.MORE:
                            case Operator.MORE_EQUAL:
                                AddWhere(connection.ConvertOperatorToString((Operator)item));
                                break;
                            case Operator.CONTAINS:
                                {
                                    if (!(where.Items[i-1] is DatabaseFieldInfo ))
                                    {
                                         throw new ArgumentException("Where is not valid. Contains syntax invalid");
                                    }

                                    if (!(where.Items[i + 1] is String) && !(where.Items[i + 1] is DatabaseFieldInfo))
                                    {
                                        throw new ArgumentException("Where is not valid. Contains syntax invalid");
                                    }
                                    String aux="";
                                    DatabaseFieldInfo info = (DatabaseFieldInfo)where.Items[i - 1];
                                    AddFrom(info.TableName);
                                    aux += info.TableName + "." + info.Attribute.Name;
                                
                                
                                    aux += " " + connection.ConvertOperatorToString((Operator)item) + " ";

                                    if (where.Items[i + 1] is DatabaseFieldInfo)
                                    {
                                        info = (DatabaseFieldInfo)where.Items[i + 1];
                                        AddFrom(info.TableName);
                                        aux += info.TableName + "." + info.Attribute.Name;
                                    }
                                    else
                                    {
                                        aux += "'%"+where.Items[i+1]+"%'";
                                    }                                
                                    AddWhere(aux);
                                    i++;
                                }
                                break;
                        }
                    }
                    else
                    {
                        AddWhere("@Parameter" + parameterCount);
                        ret.Add("@Parameter" + parameterCount, item);
                        parameterCount++;
                    }
                }
            }
            return ret;
        }

        public void AddFrom(string table)
        {
            foreach (string item in froms)
            {
                if (table == item)
                {
                    return;
                }
            }
            froms.Add(table);
        }

        public bool HasFrom()
        {
            return froms.Count > 0;
        }

        public bool hasField()
        {
            return fields.Count > 0;
        }

        public bool hasWhere()
        {
            return where.Count > 0;
        }

        #region [ Properties ]
        public OrderBy OrderBy
        {
            get;
            set;
        }
        public int LimitStart
        {
            get;
            set;
        }
        public int LimitLength
        {
            get;
            set;
        }
        #endregion
    }
}
